### PR TITLE
Multiple select

### PR DESCRIPTION
**Pull request by mmcdermott**

### CODE REVIEW

File: src/components/lib/FlowGrid/FlowGrid.jsx [hunk 1]
```diff
@@ -56,56 +58,87 @@ export default class FlowGrid extends Component{
...
+
     let direction = keycodeToDirection(e.keyCode)
     if (direction) {
       e.preventDefault()
       const size = ({columns: this._columnCount(), rows: this._rowCount()})
-      let newLocation = new DirectionToLocation(size, this.props.selected)[direction]()
+      let newLocation = new DirectionToLocation(size, this.props.selectedCell)[direction]()
       this.props.onSelectItem(newLocation)
-    } else if (e.keyCode == '17' || e.keyCode == '224' || e.keyCode == '91') {
+      this.refs[`cell-${newLocation.row}-${newLocation.column}`].decoratedComponentInstance._focus()
```
Comment (OAGr) [inline src/components/lib/FlowGrid/FlowGrid.jsx]: is this still necessary?

File: src/components/lib/FlowGrid/FlowGrid.jsx [hunk 2]
```diff
@@ -56,56 +58,87 @@ export default class FlowGrid extends Component{
...
+  _handleEndRangeSelect(corner1) {
+    const corner2 = this.props.selectedCell
+
+    if (!isLocation(corner2)) {
+      this.props.onSelectItem(corner1)
+      return
+    }
+
+    this.props.onMultipleSelect(corner1, corner2)
+  }
```
Comment (OAGr) [inline src/components/lib/FlowGrid/FlowGrid.jsx]: what if corner1 == corner2? Also, any chance this above function could be cleaned up a bit more?

File: src/components/lib/FlowGrid/FlowGrid.jsx [hunk 3]
```diff
@@ -132,6 +165,10 @@ export default class FlowGrid extends Component{
     }
   }
 
+  componentWillUnmount() {
+    this.props.onDeSelectAll()
+  }
```
Comment (OAGr) [inline src/components/lib/FlowGrid/FlowGrid.jsx]: I think this should go in the space canvas, not here.

File: src/components/lib/FlowGrid/FlowGrid.jsx (hunk 2 above)
Comment (mmcdermott) [inline src/components/lib/FlowGrid/FlowGrid.jsx]: Multi-selecting a 1x1 grid is perfectly valid; actually is used often. Right now I believe region select controls most of the styling, so should be present on singleton selections as well (and is presently). I'll add an else statement momentarily.

File: src/components/lib/FlowGrid/FlowGrid.jsx (hunk 3 above)
Comment (mmcdermott) [inline src/components/lib/FlowGrid/FlowGrid.jsx]: This was a bit of future proofing, I guess/API separation. The space canvas also de-selects all upon unmount, but I think the FlowGrid should too. If you were to use this component, and it were to unmount, then the flowgrid has disappeared from the page; I don't think in that case it is sensible to retain which flowgrid locations were selected.

File: src/components/spaces/canvas/index.js [hunk 4]
```diff
@@ -76,15 +78,21 @@ export default class Canvas extends Component{
...
+    this._handleDeSelectAll()
   }
 
   _handleSelect(location) {
     this.props.dispatch(changeSelect(location))
+    this.props.dispatch(selectRegion(location, location))
   }
 
-  _handleDeSelect() {
+  _handleMultipleSelect(corner1,corner2) {
```
Comment (OAGr) [inline src/components/spaces/canvas/index.js]: space in between var names

File: src/lib/propagation/graph-propagation.js [hunk 5]
```diff
@@ -39,7 +40,7 @@ export class GraphPropagation {
     const orderedMetricIdsAndGraphErrors = this._orderedMetricIdsAndErrors(graphFilters)
     this.orderedMetricIds = orderedMetricIdsAndGraphErrors.map(m => m.id)
     this.orderedMetricPropagations = orderedMetricIdsAndGraphErrors.map(
-      ({id, errors}) => (new MetricPropagation(id, errors, this.id))
+      ({id, errors}) => (new MetricPropagation(id, errors, this.id, !!this.onlyMetrics && !_.some(this.onlyMetrics, m => m.id === id)))
```
Comment (OAGr) [inline src/lib/propagation/graph-propagation.js]: split into two lines if possible.

File: src/modules/copied/actions.js [hunk 6]
```diff
@@ -1,43 +1,74 @@
...
 
-    const newGuesstimate = Object.assign({}, guesstimate, {metric: newItem.id})
+    const existingMetrics = spaceMetrics.filter(m => isWithinRegion(m.location, pasteRegion))
+    existingMetrics.forEach(existingMetric => {dispatch(metricActions.removeMetric(existingMetric.id))})
 
-    const existingMetric = spaceMetrics.find(m => isAtLocation(m, location))
-    if (existingMetric) {
-      dispatch(metricActions.removeMetric(existingMetric.id))
-    }
+    newMetrics.forEach((newMetric, i) => {dispatch({type: 'ADD_METRIC', item: newMetric, newGuesstimate: newGuesstimates[i]})})
```
Comment (OAGr) [inline src/modules/copied/actions.js]: would be nice for this to be shorter

Comment (mmcdermott) [inline src/modules/copied/actions.js]: Agreed, going to leave off for now though as I don't think there is significant functionality we can extract here that wouldn't be just extracting function scope and putting it higher in the file, which IMO would increase total cost of ownership.

**Comment by OAGr** [comment]
Woo! Good job

**Pull request merged**